src/leds.c

```diff
@@ -99,18 +100,18 @@ static uint32_t color_blend(uint32_t color1, uint32_t color2, float blend) {
     return RGBW(r, g, b, w);
 }
 
-// Borrowed from WLED
 static uint32_t color_wheel(uint8_t pos) {
-    pos = 255 - pos;
-    if (pos < 85) {
-        return ((uint32_t) (255 - pos * 3) << 16) | ((uint32_t) 0 << 8) | (pos * 3);
-    } else if (pos < 170) {
-        pos -= 85;
-        return ((uint32_t) 0 << 16) | ((uint32_t) (pos * 3) << 8) | (255 - pos * 3);
-    } else {
-        pos -= 170;
-        return ((uint32_t) (pos * 3) << 16) | ((uint32_t) (255 - pos * 3) << 8) | 0;
-    }
+    // normalise pos between (0, 1)
+    float normalized_pos = (float) pos / 255.0;
+
+    // smoothly interpolate using sine
+    // the sqrt is here for gamma correction which makes it appear more natural
+    float r = sqrtf(sinf((normalized_pos + 0.0 / 3.0) * TAU) * 0.5 + 0.5);
+    float b = sqrtf(sinf((normalized_pos + 1.0 / 3.0) * TAU) * 0.5 + 0.5);
+    float g = sqrtf(sinf((normalized_pos + 2.0 / 3.0) * TAU) * 0.5 + 0.5);
```

Comment:
This is much more CPU-intensive than the previous (obviously highly optimized) calculations. Trigonometric functions are especially complicated. I even had stability issues previously, which resulted in me just implementing the `cosine_progress()` function as a very simple Bhaskara approximation - IIRC the newlib-provided variants would eventually crash for high angles, and `fmodf`-ing the angle somehow still caused occasional crashes.

What I also find interesting is you `sqrtf`-ing the colors, while there's a `cgamma()` function in `led_driver.c` that does the opposite to correct the gamma :grin: 

Since the input range for this function is clearly defined and it apparently doesn't crash, the only practical concern would be how much slower this is. I want the LED code to have an absolutely minimal impact on the balancing thread, for obvious reasons.

What I think would be ideal is to approximate the result of the above function in some more effective way. I just did a quick and dirty plot here, plotting your function and approximating it with a simple quadratic function:
https://www.desmos.com/calculator/at3sw4cepm

It obviously needs some tweaking, but I think it would work just fine for our application and would remove the expensive function calls.

Reply:
Really appreciate these reviews - this is my first time writing code for a device like this, so learning lots! 

I'll give this a try!

Reply:
As far as I can tell, the input range to `sqrt(sin(x) * 0.5 + 0.5)` is `0 .. (2/3)*tau`, so I tried to create a function that mimics that input range as best I could without using any trig. I found the Taylor Series and used it to mimic the graph: https://www.desmos.com/calculator/tuxtfbi946

Having some trouble mapping this back into `color_wheel` though :thinking: perhaps I'm doing something silly, or maybe I've misunderstood something. I'll keep looking.